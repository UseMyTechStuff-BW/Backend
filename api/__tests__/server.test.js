const request = require('supertest')
const server = require('../server')
const db = require('../data/db-config')

const user1 = { username: 'joe', password: '1234', role: 'renter' }
const user2 = { username: 'billy', password: '1234', role: 'owner' }

beforeAll(async () => {
  await db.migrate.rollback()
  await db.migrate.latest()
})
beforeEach(async () => {
  await db.seed.run()
})
afterAll(async () => {
  await db.destroy()
})

it('sanity check', () => {
  expect(true).not.toBe(false)
})

describe('server.js', () => {
  it('is the correct testing environment', async () => {
    expect(process.env.NODE_ENV).toBe('testing')
  })
})

describe('[POST] /register - register new user', () => {
  it('returns a status 201 CREATED', async () => {
    const res = await request(server).post('/api/auth/register').send(user1)
    expect(res.status).toBe(201)
  })
  it('returns a newly created user', async () => {
    const res = await request(server).post('/api/auth/register').send(user1)
    expect(res.body).toMatchObject({ username: 'joe', role: 'renter' })
  })
})

describe('[POST] /login - login as user', () => {
  it('login succesful with status 200', async () => {
    await request(server).post('/api/auth/register').send(user2)
    const res = await request(server).post('/api/auth/login').send(user2)
    expect(res.status).toBe(200)
  })
  it('responds with the welcome message', async () => {
    await request(server).post('/api/auth/register').send(user2)
    const res = await request(server).post('/api/auth/login').send(user2)
    expect(res.body.message).toBe('billy is back!')
  })
})

describe('[GET] /equipment - get all equipment', () => {
  it('get equipment succesful with status 200', async () => {
    const res = await request(server).get('/api/equipment')
    expect(res.status).toBe(200)
  })
  it('responds with equipment', async () => {
    const res = await request(server).get('/api/equipment')
    expect(res.body).toMatchSnapshot()
  })
})

describe('[DEL] /equipment/:id - unable to delete unless owner token is added to header', () => {
  it('delete equipment not succesful with status 401', async () => {
    const res = await request(server).delete('/api/equipment/1')
    expect(res.status).toBe(401)
  })
})

describe('[POST] /requests - unable to post rental request unless authorized', () => {
  it('responds with 401 because not authorized', async () => {
    const res = await request(server).post('/api/requests').send({ equipment_id: 2 })
    expect(res.status).toBe(401)
  })
})

describe("[GET] /owner - unable to get owner equip unless authorized", () => {
  it("responds with 401 because not authorized", async () =>{
    const res = await request(server).get('/api/users/owner')
    expect(res.status).toBe(401)
  })
})

describe("[GET] /renter - unable to get renter equip unless authorized", () => {
  it("responds with 401 because not authorized", async () =>{
    const res = await request(server).get('/api/users/renter')
    expect(res.status).toBe(401)
  })
})