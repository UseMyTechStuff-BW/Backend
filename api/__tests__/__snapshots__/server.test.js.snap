// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GET] /equipment - get all equipment responds with equipment 1`] = `
Array [
  Object {
    "description": "like new, great for adventure",
    "id": 1,
    "imgUrl": "https://images.unsplash.com/photo-1503071866712-4af2112f609b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1650&q=80",
    "isAvailable": true,
    "name": "EVO Gimbal SS GoPro Hero Session 5",
    "owner": Object {
      "id": 2,
      "username": "Mario",
    },
  },
  Object {
    "description": "mix music with ease",
    "id": 2,
    "imgUrl": "https://images.unsplash.com/photo-1619597361832-a568b1e0555f?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1500&q=80",
    "isAvailable": false,
    "name": "Pioneer DJ",
    "owner": Object {
      "id": 2,
      "username": "Mario",
    },
  },
  Object {
    "description": "enhance every game you play",
    "id": 3,
    "imgUrl": "https://images.unsplash.com/photo-1616081118936-562f7c13d9ca?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1500&q=80",
    "isAvailable": false,
    "name": "Astro A50 Gaming Headset",
    "owner": Object {
      "id": 2,
      "username": "Mario",
    },
  },
]
`;
